### ISSUE TITLE
[Fargate] [request]: Service Connect Health Checks

### ISSUE THREAD

**Issue by muzfuz** [aws/containers-roadmap#2334]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
Service Connect does not support application health checks. This means it attempts to route traffic to containers before they're ready.

We would like Service Connect to have configurable health checks similar to ALBs, or to respect the Docker healthchecks which are configured in the task definition.

**Which service(s) is this request for?**
Fargate - specifically Service Connect options.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
We run several "big" services which have a long startup time (10 to 60 seconds). These services communicate privately using Service Connect.

We noticed that we were getting served 503s during deploys or container restarts.

After some back and forth with AWS Support we were able to establish the following sequence of events:

1. Deployment starts
1. All containers in the task start booting
1. Service Connect sidecar is marked "HEALTHY" in the task
1. Clients start receiving 503 responses
1. The main app container finishes booting and is also marked "HEALTHY".
1. 503's stop

I received the following guidance on this from AWS Support:

> Service Connect registers the task to CloudMap during the ACTIVATING stage of the task lifecycle [[1]](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-lifecycle-explanation.html), and from my testing traffic is sent to the new task as soon as the task enters RUNNING status. ... Based on my testing, it appears that Service Connect does not wait for the container to enter into "HEALTHY" status before sending traffic.

From our POV we would like one of two things to be true here.

1. Service Connect waits for the app container to be marked as "HEALTHY" by the task before routing traffic to it.
OR
2. Service Connect provides a way of configuring a health check endpoint.

The fact that it is currently simply routing traffic to a task as soon as the Envoy sidecar becomes healthy means we need to do some pretty aggressive retries in the client applications, which works to paper over the cracks but can still lead to failure.

**Are you currently working around this issue?**
Yes. A combination of aggressive retries and long Docker health checks has proven effective.

We received the following guidance from AWS Support:

> Configure a container health check in the ECS task definition with startPeriod 60. Due to the startPeriod setting, although the new task will start in UNHEALTHY, ECS does not replace the task for 60 seconds. At the same time, the old task is kept alive. Service Connect has both tasks registered in CloudMap and will send traffic to both using round-robin.

This solution "works" but is merely a sticking plaster - it can still lead to failed requests and needlessly extends deploy / restart times.

**Comment by kshivaz** [comment]
Thanks for this request. We’d like to check into the behavior you saw more thoroughly. Could you share your support case ID so we can look at your specific setup?

**Comment by muzfuz** [comment]
@kshivaz thank you for looking at this. The case ID is 171276418200173.

**Comment by kshivaz** [comment]
Thanks @muzfuz. Team will take a look at it.

**Comment by krrrr38** [comment]
case ID 171998743200404 shows CloudMap also doesn't check healthcheck result.

**Comment by bmariesan** [comment]
@kshivaz I can confirm the same is happening randomly for us during scale in/out or restarts

**Comment by kshivaz** [comment]
@bmariesan : Please open a support case with the details of your environment / configuration and error logs, so our team can look into it.

**Comment by gillesbergerp** [comment]
We face the same issue. We have multiple applications (mostly Java and JRuby) deployed that communicate via service connect. During container startup, we frequently see requests hitting a task where the application container is not ready yet

**Comment by jenademoodley** [comment]
As a way to prevent this from occurring, an additional container can be added to the task definition with a dependency on the application container to be marked as HEALTHY (this means that there must be a health check defined for the application container). The container should be marked non-essential and designed to exit.

How this works is that ECS would only transition a task to RUNNING state only if all containers in the task have started. This method prevents the task from reaching RUNNING state until the application container is healthy.

Tested this approach using a container which intentionally calls sleep for 60 seconds before starting the webserver process, and used an additional non-essential alpine container. Without the additional container, 503s are returned as expected during a deployment. With the container, no 503s are observed.

**Comment by Seiya6329** [comment]
Thanks @jenademoodley for posting the [workaround]( https://github.com/aws/containers-roadmap/issues/2334#issuecomment-2285926075)! -

We (thanks to @rishabhpar) have also validated that this workaround works and effective. Please find the workaround guideline below

### Steps to mitigate the problem:
 - Identify the [container definition](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#container_definitions) that has the long spinup time in the task definition.
 - [Add a container health check](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#container_definition_healthcheck) to the identified container definition.  Like
```
"healthCheck":{
      "command":[
         "CMD-SHELL",
         "curl -f http://localhost/ || exit 1"
      ],
      "interval":30,
      "timeout":5,
      "retries":3,
      "startPeriod":60
   }
```
This is adjustable to your preferences.
- Add a second container to the list of container definitions. This will be a dummy container designed to exit and not consume resources. It will only spinup once the main container is healthy. See the dependsOn section
```
{
   "name":"serviceconnecthold",
   "image":"public.ecr.aws/docker/library/alpine:edge",
   "cpu":0,
   "portMappings":[],
   "essential":false,
   "environment":[],
   "environmentFiles":[],
   "mountPoints":[],
   "volumesFrom":[],
   "dependsOn":[
      {
         "containerName":"<---NAME OF THE MAIN CONTAINER--->",
         "condition":"HEALTHY"
      }
   ],
   "systemControls":[]
}
```
- Submit the new task definition revision
- Update the service with the new task definition revision.

**Comment by bmariesan** [comment]
> Thanks @jenademoodley for posting the [workaround](https://github.com/aws/containers-roadmap/issues/2334#issuecomment-2285926075)! -
> 
> We (thanks to @rishabhpar) have also validated that this workaround works and effective. Please find the workaround guideline below
> 
> ### Steps to mitigate the problem:
> * Identify the [container definition](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#container_definitions) that has the long spinup time in the task definition.
> * [Add a container health check](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#container_definition_healthcheck) to the identified container definition.  Like
> 
> ```
> "healthCheck":{
>       "command":[
>          "CMD-SHELL",
>          "curl -f http://localhost/ || exit 1"
>       ],
>       "interval":30,
>       "timeout":5,
>       "retries":3,
>       "startPeriod":60
>    }
> ```
> 
> This is adjustable to your preferences.
> 
> * Add a second container to the list of container definitions. This will be a dummy container designed to exit and not consume resources. It will only spinup once the main container is healthy. See the dependsOn section
> 
> ```
> {
>    "name":"serviceconnecthold",
>    "image":"public.ecr.aws/docker/library/alpine:edge",
>    "cpu":0,
>    "portMappings":[],
>    "essential":false,
>    "environment":[],
>    "environmentFiles":[],
>    "mountPoints":[],
>    "volumesFrom":[],
>    "dependsOn":[
>       {
>          "containerName":"<---NAME OF THE MAIN CONTAINER--->",
>          "condition":"HEALTHY"
>       }
>    ],
>    "systemControls":[]
> }
> ```
> 
> * Submit the new task definition revision
> * Update the service with the new task definition revision.

I can also confirm that the workaround above works like a charm

**Comment by jamesmudd** [comment]
Also confirm this workaround works. Would be great if AWS could implement a proper fix though

**Comment by thiagoscodelerae** [comment]
@muzfuz and all, is the service-connect container being marked as "unhealthy" in your case?

I have several ECS tasks running on EC2 using ECS service connect for internal communication. Sometimes, during new deployments, the ECS service connect container linked to these tasks becomes unhealthy, preventing the deployment from succeeding. This issue doesn't occur with every deployment.

These ECS tasks are GPU-based and take some time to start. I don't have any health check configured for the task definitions.

**Comment by muzfuz** [comment]
I can confirm, this is very much still an issue cc @thiagoscodelerae .

I'll use the above mentioned workaround for now, thank you @jenademoodley

**Comment by atalwar-sephora** [comment]
Hi Team,
We are facing the same issue and it really becomes a bottleneck for critical applications. For now we are evaluating the workaround and planning to move towards use of internal ALB. However, not having this basic health check feature defeats the purpose of Service Connect. 
Hope this is resolved soon.

**Comment by augustl** [comment]
Chiming in on this one, I haven't tried the workaround mentioned above, but we get 503's from both inside (via service connect endpoints) and outside (via ELBs) the cluster during redeploys.

EDIT: Confirmed, the workaround fixes the issue for me as well!

**Comment by bharti8085** [comment]
I also faced similar issue on service connect health check. Due to this issue/limitations, I can't use service connect on my cluster. My Case ID - 174724700900335.

**Comment by saluzafa** [comment]
Facing the same issue here, can we expect something on AWS side soon instead of having to rely on a workaround please? 🙂
@jenmlinaws @kshivaz

**Comment by rishkaul** [comment]
We're happy to share that this issue has been addressed. ECS Service Connect now honors Task's health status when routing traffic to tasks, preventing premature traffic routing during task startup.

### How it works

Service Connect will wait for tasks to report a `HEALTHY` status before making them available for traffic routing. This applies to task launches during deployments, scale-out events, or task replacements. Steady-state traffic routing continues to rely on passive health-checking using outlier detection.

### Getting started

To use this feature, define container health check(s) on your essential container(s) in your task definition and update your service. If you already have container health checks configured on your essential container(s) and have deployed a service revision after `October 31, 2025`, this behavior is already active.

### What to expect

You should see a reduction in connection failures during deployments. Egress metrics from your client applications should no longer show 5xx errors that previously occurred when requests were routed to tasks still initializing.

For more details on timeout boundaries and handling of different health statuses, see the updated documentation: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-connect-concepts.html#service-connect-concepts-health-check

This feature is now available in all regions in the AWS commercial, AWS GovCloud, and China partitions.

### Questions/Issues
If you have any questions or run into issues, feel free to open a new issue or reach out through AWS Support.

**Comment by herrhound** [comment]
Closing this issue as completed – please feel free to reopen if you have any further questions.
